File: src/crypto/vaultCrypto.ts
/**
 * WebCrypto Envelope Encryption (AES-256-GCM + PBKDF2-SHA256)
 * Mirrors the cryptographic architecture of AtomicVault:
 * - KEK (Key Encryption Key) derived from master password + salt
 * - DEK (Data Encryption Key) 256-bit AES key stored wrapped
 * - Sealed payload: IV (12 bytes) + ciphertext + tag
 * - Backup encoding with 'ATVB' magic header matching BackupCodec
 */

export interface VaultEnvelope {
  saltBase64: string;
  wrappedDek: string; // Base64 [12-byte IV + ciphertext]
  kdfParams: {
    algorithm: string;
    iterations: number;
    hash: string;
  };
}

const KDF_ITERATIONS = 100000;
const BACKUP_MAGIC = new Uint8Array([0x41, 0x54, 0x56, 0x42]); // "ATVB"

export class VaultCrypto {
  /**
   * Derives a KEK from password and Base64 salt using PBKDF2
   */
  static async deriveKek(password: string, saltBytes: Uint8Array): Promise<CryptoKey> {
    const enc = new TextEncoder();
    const passKey = await crypto.subtle.importKey(
      'raw',
      enc.encode(password),
      { name: 'PBKDF2' },
      false,
      ['deriveKey']
    );

    return crypto.subtle.deriveKey(
      {
        name: 'PBKDF2',
        salt: saltBytes as BufferSource,
        iterations: KDF_ITERATIONS,
        hash: 'SHA-256',
      },
      passKey,
      { name: 'AES-GCM', length: 256 },
      true,
      ['encrypt', 'decrypt', 'wrapKey', 'unwrapKey']
    );
  }

  /**
   * Generates a 256-bit DEK
   */
  static async generateDek(): Promise<CryptoKey> {
    return crypto.subtle.generateKey(
      { name: 'AES-GCM', length: 256 },
      true,
      ['encrypt', 'decrypt']
    );
  }

  /**
   * Wraps DEK using KEK
   */
  static async wrapDek(kek: CryptoKey, dek: CryptoKey): Promise<string> {
    const rawDek = await crypto.subtle.exportKey('raw', dek);
    const iv = crypto.getRandomValues(new Uint8Array(12));
    const encrypted = await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv },
      kek,
      rawDek
    );

    const combined = new Uint8Array(iv.length + encrypted.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(encrypted), iv.length);
    return this.bytesToBase64(combined);
  }

  /**
   * Unwraps DEK using KEK
   */
  static async unwrapDek(kek: CryptoKey, wrappedDekBase64: string): Promise<CryptoKey> {
    const combined = this.base64ToBytes(wrappedDekBase64);
    if (combined.length < 12 + 16) {
      throw new Error('Invalid wrapped DEK length');
    }
    const iv = combined.slice(0, 12);
    const ciphertext = combined.slice(12);

    const rawDek = await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv },
      kek,
      ciphertext
    );

    return crypto.subtle.importKey(
      'raw',
      rawDek,
      { name: 'AES-GCM', length: 256 },
      true,
      ['encrypt', 'decrypt']
    );
  }

  /**
   * Encrypts plaintext string using DEK
   */
  static async seal(dek: CryptoKey, plaintext: string): Promise<string> {
    const enc = new TextEncoder();
    const data = enc.encode(plaintext);
    const iv = crypto.getRandomValues(new Uint8Array(12));
    const encrypted = await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv },
      dek,
      data
    );

    const combined = new Uint8Array(iv.length + encrypted.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(encrypted), iv.length);
    return this.bytesToBase64(combined);
  }

  /**
   * Decrypts ciphertext string using DEK
   */
  static async open(dek: CryptoKey, ciphertextBase64: string): Promise<string> {
    const combined = this.base64ToBytes(ciphertextBase64);
    if (combined.length < 12 + 16) {
      throw new Error('Invalid ciphertext payload');
    }
    const iv = combined.slice(0, 12);
    const ciphertext = combined.slice(12);

    const decrypted = await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv },
      dek,
      ciphertext
    );

    const dec = new TextDecoder();
    return dec.decode(decrypted);
  }

  /**
   * Export encrypted backup with passphrase (ATVB magic envelope)
   */
  static async exportBackup(jsonData: string, passphrase: string): Promise<Uint8Array> {
    const rawSalt = crypto.getRandomValues(new Uint8Array(16));
    const backupKey = await this.deriveKek(passphrase, rawSalt);

    const enc = new TextEncoder();
    const jsonBytes = enc.encode(jsonData);
    const iv = crypto.getRandomValues(new Uint8Array(12));
    const encrypted = await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv },
      backupKey,
      jsonBytes
    );

    // Envelope: [MAGIC 4B][1B salt len][salt bytes][IV 12B][encrypted data]
    const encryptedBlob = new Uint8Array(iv.length + encrypted.byteLength);
    encryptedBlob.set(iv, 0);
    encryptedBlob.set(new Uint8Array(encrypted), iv.length);

    const totalLen = BACKUP_MAGIC.length + 1 + rawSalt.length + encryptedBlob.length;
    const result = new Uint8Array(totalLen);
    let offset = 0;

    result.set(BACKUP_MAGIC, offset);
    offset += BACKUP_MAGIC.length;

    result[offset] = rawSalt.length;
    offset += 1;

    result.set(rawSalt, offset);
    offset += rawSalt.length;

    result.set(encryptedBlob, offset);
    return result;
  }

  /**
   * Import encrypted backup with passphrase
   */
  static async importBackup(backupBytes: Uint8Array, passphrase: string): Promise<string> {
    if (backupBytes.length < BACKUP_MAGIC.length + 1 + 16 + 28) {
      throw new Error('Backup file is too small or corrupted');
    }

    for (let i = 0; i < BACKUP_MAGIC.length; i++) {
      if (backupBytes[i] !== BACKUP_MAGIC[i]) {
        throw new Error('Invalid backup file format (magic mismatch)');
      }
    }

    let offset = BACKUP_MAGIC.length;
    const saltLen = backupBytes[offset];
    offset += 1;

    if (saltLen !== 16) {
      throw new Error(`Unsupported salt length: ${saltLen}`);
    }

    const salt = backupBytes.slice(offset, offset + saltLen);
    offset += saltLen;

    const encryptedBlob = backupBytes.slice(offset);
    if (encryptedBlob.length < 12 + 16) {
      throw new Error('Encrypted payload too short');
    }

    const iv = encryptedBlob.slice(0, 12);
    const ciphertext = encryptedBlob.slice(12);

    const backupKey = await this.deriveKek(passphrase, salt);
    try {
      const decrypted = await crypto.subtle.decrypt(
        { name: 'AES-GCM', iv },
        backupKey,
        ciphertext
      );
      const dec = new TextDecoder();
      return dec.decode(decrypted);
    } catch {
      throw new Error('Incorrect backup passphrase or corrupted file');
    }
  }

  /**
   * Compute SHA-256 hex string
   */
  static async sha256(input: string): Promise<string> {
    const enc = new TextEncoder();
    const data = enc.encode(input);
    const hash = await crypto.subtle.digest('SHA-256', data);
    return Array.from(new Uint8Array(hash))
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
  }

  /**
   * Compute HMAC-SHA-256 hex string
   */
  static async hmacSha256(keyBytes: Uint8Array, input: string): Promise<string> {
    const cryptoKey = await crypto.subtle.importKey(
      'raw',
      keyBytes as BufferSource,
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign']
    );
    const enc = new TextEncoder();
    const sig = await crypto.subtle.sign('HMAC', cryptoKey, enc.encode(input));
    return Array.from(new Uint8Array(sig))
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
  }

  static bytesToBase64(bytes: Uint8Array): string {
    let binary = '';
    for (let i = 0; i < bytes.byteLength; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    return btoa(binary);
  }

  static base64ToBytes(base64: string): Uint8Array {
    const binary = atob(base64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }
}

File: app/src/main/java/com/atomicvault/android/crypto/VaultCrypto.kt
package com.atomicvault.android.crypto

import android.util.Base64
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

object VaultCrypto {

    private const val KDF_ITERATIONS = 100000
    private const val KEY_LENGTH_BITS = 256
    private const val GCM_IV_LENGTH_BYTES = 12
    private const val GCM_TAG_LENGTH_BITS = 128
    private val BACKUP_MAGIC = byteArrayOf(0x41, 0x54, 0x56, 0x42) // "ATVB"

    private val secureRandom = SecureRandom()

    fun generateSalt(): ByteArray {
        val salt = ByteArray(16)
        secureRandom.nextBytes(salt)
        return salt
    }

    fun deriveKek(password: String, salt: ByteArray): SecretKey {
        val spec = PBEKeySpec(password.toCharArray(), salt, KDF_ITERATIONS, KEY_LENGTH_BITS)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val keyBytes = factory.generateSecret(spec).encoded
        return SecretKeySpec(keyBytes, "AES")
    }

    fun generateDek(): SecretKey {
        val keyBytes = ByteArray(32) // 256 bits
        secureRandom.nextBytes(keyBytes)
        return SecretKeySpec(keyBytes, "AES")
    }

    fun wrapDek(kek: SecretKey, dek: SecretKey): String {
        val iv = ByteArray(GCM_IV_LENGTH_BYTES)
        secureRandom.nextBytes(iv)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, kek, GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))
        val encrypted = cipher.doFinal(dek.encoded)

        val combined = ByteArray(iv.size + encrypted.size)
        System.arraycopy(iv, 0, combined, 0, iv.size)
        System.arraycopy(encrypted, 0, combined, iv.size, encrypted.size)
        return Base64.encodeToString(combined, Base64.NO_WRAP)
    }

    fun unwrapDek(kek: SecretKey, wrappedDekBase64: String): SecretKey {
        val combined = Base64.decode(wrappedDekBase64, Base64.NO_WRAP)
        if (combined.size < GCM_IV_LENGTH_BYTES + 16) {
            throw IllegalArgumentException("Invalid wrapped DEK length")
        }
        val iv = combined.copyOfRange(0, GCM_IV_LENGTH_BYTES)
        val ciphertext = combined.copyOfRange(GCM_IV_LENGTH_BYTES, combined.size)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, kek, GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))
        val rawDek = cipher.doFinal(ciphertext)
        return SecretKeySpec(rawDek, "AES")
    }

    fun seal(dek: SecretKey, plaintext: String): String {
        val data = plaintext.toByteArray(StandardCharsets.UTF_8)
        val iv = ByteArray(GCM_IV_LENGTH_BYTES)
        secureRandom.nextBytes(iv)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, dek, GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))
        val encrypted = cipher.doFinal(data)

        val combined = ByteArray(iv.size + encrypted.size)
        System.arraycopy(iv, 0, combined, 0, iv.size)
        System.arraycopy(encrypted, 0, combined, iv.size, encrypted.size)
        return Base64.encodeToString(combined, Base64.NO_WRAP)
    }

    fun open(dek: SecretKey, ciphertextBase64: String): String {
        val combined = Base64.decode(ciphertextBase64, Base64.NO_WRAP)
        if (combined.size < GCM_IV_LENGTH_BYTES + 16) {
            throw IllegalArgumentException("Invalid ciphertext payload")
        }
        val iv = combined.copyOfRange(0, GCM_IV_LENGTH_BYTES)
        val ciphertext = combined.copyOfRange(GCM_IV_LENGTH_BYTES, combined.size)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, dek, GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))
        val decrypted = cipher.doFinal(ciphertext)
        return String(decrypted, StandardCharsets.UTF_8)
    }

    fun exportBackup(jsonData: String, passphrase: String): ByteArray {
        val salt = generateSalt()
        val backupKey = deriveKek(passphrase, salt)

        val iv = ByteArray(GCM_IV_LENGTH_BYTES)
        secureRandom.nextBytes(iv)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, backupKey, GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))
        val encrypted = cipher.doFinal(jsonData.toByteArray(StandardCharsets.UTF_8))

        val encryptedBlob = ByteArray(iv.size + encrypted.size)
        System.arraycopy(iv, 0, encryptedBlob, 0, iv.size)
        System.arraycopy(encrypted, 0, encryptedBlob, iv.size, encrypted.size)

        val totalLen = BACKUP_MAGIC.size + 1 + salt.size + encryptedBlob.size
        val result = ByteArray(totalLen)
        var offset = 0

        System.arraycopy(BACKUP_MAGIC, 0, result, offset, BACKUP_MAGIC.size)
        offset += BACKUP_MAGIC.size

        result[offset] = salt.size.toByte()
        offset += 1

        System.arraycopy(salt, 0, result, offset, salt.size)
        offset += salt.size

        System.arraycopy(encryptedBlob, 0, result, offset, encryptedBlob.size)
        return result
    }

    fun importBackup(backupBytes: ByteArray, passphrase: String): String {
        if (backupBytes.size < BACKUP_MAGIC.size + 1 + 16 + 28) {
            throw IllegalArgumentException("Backup file is too small or corrupted")
        }

        for (i in BACKUP_MAGIC.indices) {
            if (backupBytes[i] != BACKUP_MAGIC[i]) {
                throw IllegalArgumentException("Invalid backup file format (magic mismatch)")
            }
        }

        var offset = BACKUP_MAGIC.size
        val saltLen = backupBytes[offset].toInt() and 0xFF
        offset += 1

        if (saltLen != 16) {
            throw IllegalArgumentException("Unsupported salt length: $saltLen")
        }

        val salt = backupBytes.copyOfRange(offset, offset + saltLen)
        offset += saltLen

        val encryptedBlob = backupBytes.copyOfRange(offset, backupBytes.size)
        if (encryptedBlob.size < GCM_IV_LENGTH_BYTES + 16) {
            throw IllegalArgumentException("Encrypted payload too short")
        }

        val iv = encryptedBlob.copyOfRange(0, GCM_IV_LENGTH_BYTES)
        val ciphertext = encryptedBlob.copyOfRange(GCM_IV_LENGTH_BYTES, encryptedBlob.size)

        val backupKey = deriveKek(passphrase, salt)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, backupKey, GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))
        val decrypted = cipher.doFinal(ciphertext)
        return String(decrypted, StandardCharsets.UTF_8)
    }

    fun sha256(input: String): String {
        val md = MessageDigest.getInstance("SHA-256")
        val digest = md.digest(input.toByteArray(StandardCharsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    fun hmacSha256(keyBytes: ByteArray, input: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        val secretKey = SecretKeySpec(keyBytes, "HmacSHA256")
        mac.init(secretKey)
        val sig = mac.doFinal(input.toByteArray(StandardCharsets.UTF_8))
        return sig.joinToString("") { "%02x".format(it) }
    }

    fun generateTotp(secret: String, timeStepSeconds: Long = 30): String {
        if (secret.isBlank()) return "------"
        val keyBytes = decodeBase32(secret) ?: return "INVALID"
        val epochSeconds = System.currentTimeMillis() / 1000
        val counter = epochSeconds / timeStepSeconds

        val buffer = ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN)
        buffer.putLong(counter)
        val data = buffer.array()

        val mac = Mac.getInstance("HmacSHA1")
        mac.init(SecretKeySpec(keyBytes, "HmacSHA1"))
        val hmac = mac.doFinal(data)

        val offset = hmac[hmac.size - 1].toInt() and 0x0F
        val binary = ((hmac[offset].toInt() and 0x7F) shl 24) or
                ((hmac[offset + 1].toInt() and 0xFF) shl 16) or
                ((hmac[offset + 2].toInt() and 0xFF) shl 8) or
                (hmac[offset + 3].toInt() and 0xFF)

        val otp = binary % 1000000
        return String.format("%06d", otp)
    }

    fun getTotpSecondsRemaining(timeStepSeconds: Long = 30): Int {
        val epochSeconds = System.currentTimeMillis() / 1000
        return (timeStepSeconds - (epochSeconds % timeStepSeconds)).toInt()
    }

    private fun decodeBase32(base32: String): ByteArray? {
        val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        val clean = base32.uppercase().replace(Regex("[^A-Z2-7]"), "")
        if (clean.isEmpty()) return null
        var bits = 0
        var value = 0
        val output = mutableListOf<Byte>()

        for (c in clean) {
            val idx = alphabet.indexOf(c)
            if (idx == -1) return null
            value = (value shl 5) or idx
            bits += 5
            if (bits >= 8) {
                output.add(((value ushr (bits - 8)) and 0xFF).toByte())
                bits -= 8
            }
        }
        return output.toByteArray()
    }
}
